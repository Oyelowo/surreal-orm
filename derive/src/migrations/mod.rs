pub mod table;

pub use table::*;
