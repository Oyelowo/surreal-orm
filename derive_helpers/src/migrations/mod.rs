/*
 * Author: Oyelowo Oyedayo
 * Email: oyelowo.oss@gmail.com
 * Copyright (c) 2024 Oyelowo Oyedayo
 * Licensed under the MIT license
 */

pub mod table;
