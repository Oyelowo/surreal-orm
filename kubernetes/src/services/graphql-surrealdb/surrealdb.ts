import { ServiceDeployment } from '../../shared/deployment.js';
import { AppConfigs, AppEnvVars, TikVDbEnvVars } from '../../types/ownTypes.js';
import { getIngressUrl } from '../../infrastructure/ingress/hosts.js';
import { PlainSecretsManager } from '../../../scripts/utils/plainSecretsManager.js';
import { getEnvVarsForKubeManifests } from '../../shared/environmentVariablesForManifests.js';
import { graphqlSurrealdb } from './app.js';

const env = getEnvVarsForKubeManifests();

const secrets = new PlainSecretsManager('services', 'graphql-surrealdb', 'local').getSecrets();

type SurrealDbEnvVars = AppEnvVars & TikVDbEnvVars<'applications'>;

const surrealDbEnvVars: SurrealDbEnvVars = {
    APP_ENVIRONMENT: env.ENVIRONMENT,
    APP_HOST: '0.0.0.0',
    APP_PORT: '8000',
    APP_EXTERNAL_BASE_URL: getIngressUrl({ environment: env.ENVIRONMENT }),
    TIKV_HOST: 'tikv-pd.applications',
    TIKV_NAME: 'tikv',
    TIKV_PORT: '2379',
    TIKV_SERVICE_NAME: 'tikv',
    TIKV_STORAGE_CLASS: 'linode-block-storage-retain',
};

// Surrealdb is a compute/logic protocol layer using TiKV as persistent layer
// so, we can also deploy it as a kubernetes deployment or statefulset
export const surrealdbSettings: AppConfigs<'surrealdb', 'applications', SurrealDbEnvVars> = {
    kubeConfig: {
        requestMemory: '70Mi',
        requestCpu: '100m',
        limitMemory: '200Mi',
        limitCpu: '100m',
        replicaCount: 2,
        readinessProbePort: 8000,
        host: '0.0.0.0',
        image: `surrealdb/surrealdb:1.0.0-beta.8`,
        command: ['/surreal'],
        // TODO: Change hardcoded credentials once surrealdb is 1.0
        commandArgs: [
            'start',
            '--log',
            'debug',
            '--user',
            secrets.SURREALDB_ROOT_USERNAME,
            '--pass',
            secrets.SURREALDB_ROOT_PASSWORD,
            `tikv://${surrealDbEnvVars.TIKV_HOST}:${surrealDbEnvVars.TIKV_PORT}`,
        ],
    },
    envVars: surrealDbEnvVars,
    metadata: {
        name: 'surrealdb',
        namespace: 'applications',
    },
};

export const surrealDbDeployment = new ServiceDeployment('surrealdb', surrealdbSettings);
surrealDbDeployment.setProvider(graphqlSurrealdb.getProvider());
