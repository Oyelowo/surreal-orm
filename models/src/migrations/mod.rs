/*
 * Author: Oyelowo Oyedayo
 * Email: oyelowo.oss@gmail.com
 * Copyright (c) 2023 Oyelowo Oyedayo
 * Licensed under the MIT license
 */

use chrono::Utc;
use serde::{Deserialize, Serialize};
use surreal_orm::{
    arr, cond, create_table_resources,
    statements::{define_event, define_index, select},
    *,
};

#[derive(Debug, Clone)]
pub struct Resources;

#[derive(Debug, Clone)]
pub struct ResourcesV2;

impl DbResources for ResourcesV2 {
    create_table_resources!(AnimalV2, Crop, AnimalEatsCropV2, PlanetV2, NewStuff);
}
#[derive(Node, Serialize, Deserialize, Debug, Clone, Default)]
#[serde(rename_all = "camelCase")]
#[surreal_orm(table_name = "new_stuff", schemafull)]
pub struct NewStuff {
    // Test renaming tomorrow
    pub id: SurrealSimpleId<Self>,
    pub first_name: String,
    pub created_at: chrono::DateTime<Utc>,
    pub updated_at: chrono::DateTime<Utc>,
}

impl TableResources for NewStuff {}

impl DbResources for Resources {
    create_table_resources!(Animal, Crop, AnimalEatsCrop, Student, Planet);
}

#[derive(Node, Serialize, Deserialize, Debug, Clone, Default)]
#[serde(rename_all = "camelCase")]
#[surreal_orm(table_name = "planet", schemafull)]
pub struct Planet {
    // Test renaming tomorrow
    pub id: SurrealSimpleId<Self>,
    pub first_name: String,
    pub population: u64,
    pub created_at: chrono::DateTime<Utc>,
    pub updated_at: chrono::DateTime<Utc>,
    pub labels: Vec<String>,
}

impl TableResources for Planet {}

#[derive(Node, Serialize, Deserialize, Debug, Clone, Default)]
#[serde(rename_all = "camelCase")]
#[surreal_orm(table_name = "planet", schemafull, relax_table_name)]
pub struct PlanetV2 {
    // Test renaming tomorrow
    pub id: SurrealSimpleId<Self>,
    #[surreal_orm(old_name = "firstName")]
    pub new_name: String,
    pub population: u64,
    pub created_at: chrono::DateTime<Utc>,
    pub updated_at: chrono::DateTime<Utc>,
    pub labels: Vec<String>,
}

impl TableResources for PlanetV2 {}

#[derive(Node, Serialize, Deserialize, Debug, Clone, Default)]
#[serde(rename_all = "camelCase")]
#[surreal_orm(table_name = "student", schemafull)]
pub struct Student {
    pub id: SurrealSimpleId<Self>,
    pub university: String,
    pub age: u8,
    pub updated_at: chrono::DateTime<Utc>,
    pub created_at: chrono::DateTime<Utc>,
}

impl TableResources for Student {}

#[derive(Node, Serialize, Deserialize, Debug, Clone, Default)]
#[serde(rename_all = "camelCase")]
#[surreal_orm(table_name = "animal", schemafull)]
pub struct Animal {
    pub id: SurrealSimpleId<Self>,
    pub species: String,
    // Improve error essage for old_nmae using word similarity algo
    pub attributes: Vec<String>,
    pub created_at: chrono::DateTime<Utc>,
    pub updated_at: chrono::DateTime<Utc>,
    pub velocity: u64,
}

impl TableResources for Animal {
    fn events_definitions() -> Vec<Raw> {
        let animal::Schema {
            species, velocity, ..
        } = Self::schema();

        let event1 = define_event("event1".to_string())
            .on_table("animal".to_string())
            .when(cond(species.eq("Homo Erectus").and(velocity.gt(545))))
            .then(select(All).from(Crop::table_name()))
            .to_raw();

        let event2 = define_event("event2".to_string())
            .on_table("animal".to_string())
            .when(cond(species.eq("Homo Sapien")).and(velocity.lt(10)))
            .then(select(All).from(AnimalEatsCrop::table_name()))
            .to_raw();
        vec![event1, event2]
    }

    fn indexes_definitions() -> Vec<Raw> {
        let animal::Schema {
            species, velocity, ..
        } = Self::schema();

        let idx1 = define_index("species_speed_idx".to_string())
            .on_table(Self::table_name())
            .fields(arr![species, velocity])
            .unique()
            .to_raw();

        vec![idx1]
    }
}

#[derive(Node, Serialize, Deserialize, Debug, Clone, Default)]
#[serde(rename_all = "camelCase")]
#[surreal_orm(table_name = "animal", schemafull, relax_table_name)]
pub struct AnimalV2 {
    pub id: SurrealSimpleId<Self>,
    pub species: String,
    #[surreal_orm(old_name = "attributes")]
    pub characteristics: Vec<String>,
    pub velocity: u64,
}

impl TableResources for AnimalV2 {
    fn events_definitions() -> Vec<Raw> {
        let animal_v_2::Schema {
            species, velocity, ..
        } = Self::schema();

        let event1 = define_event("event1".to_string())
            .on_table("animal".to_string())
            .when(cond(species.eq("Homo Habillis").and(velocity.gt(545))))
            .then(select(All).from(Crop::table_name()))
            .to_raw();

        vec![event1]
    }

    fn indexes_definitions() -> Vec<Raw> {
        let animal_v_2::Schema {
            characteristics,
            velocity,
            ..
        } = Self::schema();

        let idx1 = define_index("species_speed_idx".to_string())
            .on_table(Self::table_name())
            .fields(arr![velocity, characteristics])
            .unique()
            .to_raw();

        vec![idx1]
    }
}

#[derive(Edge, Serialize, Deserialize, Debug, Clone, Default)]
#[serde(rename_all = "camelCase")]
#[surreal_orm(table_name = "eats", schemafull)]
pub struct Eats<In: Node, Out: Node> {
    pub id: SurrealSimpleId<Self>,
    #[serde(rename = "in")]
    pub in_: In,
    pub out: Out,
    pub place: String,
    pub created_at: chrono::DateTime<Utc>,
}

pub type AnimalEatsCrop = Eats<Animal, Crop>;
impl TableResources for AnimalEatsCrop {}

#[derive(Edge, Serialize, Deserialize, Debug, Clone, Default)]
#[serde(rename_all = "camelCase")]
#[surreal_orm(table_name = "eats", schemafull, relax_table_name)]
pub struct EatsV2<In: Node, Out: Node> {
    pub id: SurrealSimpleId<Self>,
    #[serde(rename = "in")]
    pub in_: In,
    pub out: Out,
    pub location: String,
    pub created_at: chrono::DateTime<Utc>,
    pub updated_at: chrono::DateTime<Utc>,
}

pub type AnimalEatsCropV2 = EatsV2<AnimalV2, Crop>;
impl TableResources for AnimalEatsCropV2 {}

#[derive(Node, Serialize, Deserialize, Debug, Clone, Default)]
#[serde(rename_all = "camelCase")]
#[surreal_orm(table_name = "crop", schemafull)]
pub struct Crop {
    pub id: SurrealSimpleId<Self>,
    pub color: String,
}

impl TableResources for Crop {}
