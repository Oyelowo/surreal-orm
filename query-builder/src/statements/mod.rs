/*utils_chain
 * Author: Oyelowo Oyedayo
 * Email: oyelowo.oss@gmail.com
 * Copyright (c) 2023 Oyelowo Oyedayo
 * Licensed under the MIT license
 */

use std::fmt::Display;

pub(crate) mod break_;
pub(crate) mod continue_;
pub(crate) mod create;
pub(crate) mod create_v2;
pub(crate) mod define_analyzer;
pub(crate) mod define_database;
pub(crate) mod define_event;
pub(crate) mod define_field;
pub(crate) mod define_function;
pub(crate) mod define_index;
pub(crate) mod define_model;
pub(crate) mod define_namespace;
pub(crate) mod define_param;
pub(crate) mod define_scope;
pub(crate) mod define_table;
pub(crate) mod define_token;
pub(crate) mod define_user;
pub(crate) mod delete;
pub(crate) mod for_loop;
pub(crate) mod for_permission;
pub(crate) mod ifelse;
pub(crate) mod info;
pub(crate) mod insert;
pub(crate) mod let_;
pub(crate) mod live_select;
pub(crate) mod relate;
pub(crate) mod remove_analyzer;
pub(crate) mod remove_database;
pub(crate) mod remove_event;
pub(crate) mod remove_field;
pub(crate) mod remove_function;
pub(crate) mod remove_index;
pub(crate) mod remove_login;
pub(crate) mod remove_model;
pub(crate) mod remove_namespace;
pub(crate) mod remove_param;
pub(crate) mod remove_scope;
pub(crate) mod remove_table;
pub(crate) mod remove_token;
pub(crate) mod remove_user;
pub(crate) mod return_;
pub(crate) mod select;
pub(crate) mod select_surreal_id_range;
pub(crate) mod show;
pub(crate) mod sleep;
pub(crate) mod subquery;
pub(crate) mod throw;
pub(crate) mod transaction;
pub(crate) mod update;
pub(crate) mod use_;
pub(crate) mod utils_block;
pub(crate) mod utils_chain;

pub use break_::{break_, BreakStatement};
pub use continue_::{continue_, ContinueStatement};
pub use create::{create, create_only, CreateStatement};
pub use create_v2::{create_v2, CreateStatementV2};
pub use define_analyzer::{
    define_analyzer, AnalyzerFilter, DefineAnalyzerStatement, SnowballLanguage, Tokenizer,
};
pub use define_database::{define_database, DefineDatabaseStatement};
pub use define_event::{define_event, DefineEventStatement};
pub use define_field::{define_field, DefineFieldStatement};
pub use define_function::{define_function, DefineFunctionStatement, FunctionArgument};
pub use define_index::{define_index, search_analyzer, DefineIndexStatement, Scoring};
pub use define_model::{define_model, DefineModelStatement, ModelName, ModelVersion};
pub use define_namespace::{define_namespace, DefineNamespaceStatement};
pub use define_param::{define_param, DefineParamStatement};
pub use define_scope::{define_scope, DefineScopeStatement};
pub use define_table::{define_table, DefineTableStatement};
pub use define_token::{define_token, DefineTokenStatement};
pub use define_user::{define_user, DefineUserStatement, UserRole};
pub use delete::{delete, delete_only, DeleteStatement, DeleteStatementMini};
pub use for_loop::{for_, ForLoopStatement};
pub use for_permission::{for_permission, ForPermission, Permissions};
pub use ifelse::{if_, IfElseStatement};
pub use info::{info_for, InfoStatement};
pub use insert::{insert, InsertStatement, Insertables};
pub use let_::{let_, LetStatement};
pub use live_select::{live, LiveSelectStatement};
pub use relate::{relate, relate_only, RelateStatement};
pub use remove_analyzer::{remove_analyzer, RemoveAnalyzerStatement};
pub use remove_database::{remove_database, RemoveDatabaseStatement};
pub use remove_event::{remove_event, RemoveEventStatement};
pub use remove_field::{remove_field, RemoveFieldStatement};
pub use remove_function::{remove_function, RemoveFunctionStatement};
pub use remove_index::{remove_index, RemoveIndexStatement};
pub use remove_login::{remove_login, RemoveLoginStatement};
pub use remove_model::{remove_model, RemoveModelStatement};
pub use remove_namespace::{remove_namespace, RemoveNamespaceStatement};
pub use remove_param::{remove_param, RemoveParamStatement};
pub use remove_scope::{remove_scope, RemoveScopeStatement};
pub use remove_table::{remove_table, RemoveTableStatement};
pub use remove_token::{remove_token, RemoveTokenStatement};
pub use remove_user::{remove_user, RemoveUserStatement, UserPermissionScope};
pub use return_::{return_, ReturnStatement};
pub use select::{
    order, select, select_diff, select_value, Order, Orderables, SelectStatement,
    SelectStatementMini, Selectables, Splittables,
};
pub use show::{show_changes_for_table, ShowChangesStatement};
pub use sleep::{sleep, SleepStatement};
pub use subquery::Subquery;
pub use throw::{throw, ThrowStatement};
pub use transaction::{
    begin_transaction, transaction_deprecated, BeginTransactionStatement, TransactionCompletion,
};
pub use update::{update, update_only, UpdateStatement};
pub use use_::{use_, UseStatement};

/// helpers for statements
pub mod utils {
    pub use super::utils_block::{block, block_deprecated, Block};
    pub use super::utils_chain::{chain, Chainable, QueryChain};
}

#[allow(missing_docs)]
pub enum NamespaceOrDatabase {
    Namespace,
    Database,
}

impl Display for NamespaceOrDatabase {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let stringified = match self {
            NamespaceOrDatabase::Namespace => "NAMESPACE",
            NamespaceOrDatabase::Database => "DATABASE",
        };
        write!(f, "{}", stringified)
    }
}
