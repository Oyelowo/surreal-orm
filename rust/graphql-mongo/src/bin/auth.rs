use tracing;
use tracing_subscriber;

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    // let p = store_password(credentials.clone()).unwrap();
    let pass = spawn_blocking_with_tracing(move || {
        create_password_hash(PlainPassword::new("Oyelowo".into()))
    })
    .await?
    .context("Faailed to hash passwor")?;

    // let pass = create_password_hash(PlainPassword::new("Oyelowo".into()))?;
    println!("HGRGHJG: {:?}", pass.0.expose_secret());

    validate_credentials(PlainPassword::new("Oyelowo".into()), pass)
        .await
        .unwrap();

    Ok(())
}

pub struct PlainPassword(Secret<String>);
impl PlainPassword {
    fn new(pass: String) -> Self {
        Self(Secret::new(pass))
    }

    fn to_bytes(&self) -> &[u8] {
        self.0.expose_secret().as_bytes()
    }
}

// impl From<PlainPassword> for String {
//     fn from(p: PlainPassword) -> String {
//         let k = p.0.expose_secret().to_owned();
//         k
//     }
// }

pub struct PasswordHashPHC(Secret<String>);
impl PasswordHashPHC {
    fn new(pass: String) -> Self {
        Self(Secret::new(pass))
    }

    fn as_str(&self) -> &str {
        self.0.expose_secret().as_str()
    }
}

// impl From<PasswordHashPHC> for &str {
//     fn from(_: PasswordHashPHC) -> Self {
//         todo!()
//     }
// }
use anyhow::Context;
use argon2::{Algorithm, Argon2, Params, PasswordHash, PasswordVerifier, Version};
use secrecy::{ExposeSecret, Secret};
use thiserror;

//  return a 500 for UnexpectedError, while AuthErrors should result into a 401.
#[derive(thiserror::Error, Debug)]
pub enum AuthError {
    #[error("Authentication failed.")]
    InvalidCredentials(#[source] anyhow::Error),

    #[error(transparent)]
    UnexpectedError(#[from] anyhow::Error),
}

use tokio::task::JoinHandle;

pub fn spawn_blocking_with_tracing<F, R>(f: F) -> JoinHandle<R>
where
    F: FnOnce() -> R + Send + 'static,
    R: Send + 'static,
{
    let current_span = tracing::Span::current();
    tokio::task::spawn_blocking(move || current_span.in_scope(f))
}

// #[tracing::instrument(name = "Validate credentials", skip(credentials, pool))]
// #[tracing::instrument(
//     name = "Verify password hash",
//     skip(expected_password_hash, password_candidate)
// )]
async fn validate_credentials(
    plain_password: PlainPassword,
    expected_password_hash: PasswordHashPHC,
) -> anyhow::Result<bool, AuthError> {
    // This executes before spawning the new thread
    spawn_blocking_with_tracing(move || {
        verify_password_hash(plain_password, expected_password_hash)
    })
    .await
    .context("Failed to spawn blocking task.")
    .map_err(AuthError::UnexpectedError)??;

    println!("Yaay! Password successfully validated!!!!");
    // let password_hash = hasher.hash_password_into(credentials.password.expose_secret().as_bytes(), salt, out)
    Ok(true)
}

#[tracing::instrument(
    name = "Verify password hash",
    skip(expected_password_hash, password_candidate)
)]
fn verify_password_hash(
    password_candidate: PlainPassword,
    expected_password_hash: PasswordHashPHC,
) -> Result<(), AuthError> {
    /*
    PHC String Format
    # ${algorithm}${algorithm version}${$-separated algorithm parameters}${hash}${salt}
    $argon2id$v=19$m=65536,t=2,p=1$gZiV/M1gPc22ElAH/Jh1Hw$CWOrkoo7oJBQ/iyh7uJ0LO2aLEfrHwTWllSAxT0zRno
    */
    let expected_password_hash = PasswordHash::new(expected_password_hash.as_str())
        .context("Failed to parse hash in PHC string format.")
        .map_err(AuthError::UnexpectedError)?;

    Argon2::default()
        .verify_password(password_candidate.to_bytes(), &expected_password_hash)
        .context("Invalid password.")
        .map_err(AuthError::InvalidCredentials)
}

use argon2::{password_hash::SaltString, PasswordHasher};

fn create_password_hash<'a>(password: PlainPassword) -> anyhow::Result<PasswordHashPHC, AuthError> {
    let salt = SaltString::generate(&mut rand::thread_rng());
    // We don't care about the exact Argon2 parameters here
    // given that it's for testing purposes!
    // let password_hash = Argon2::default()
    //     .hash_password("pass".as_bytes(), &salt)
    //     .unwrap()
    //     .to_string();
    // println!("password_hash11111: {:?}", password_hash);

    let params = Params::new(15000, 2, 1, None).context("Error building Argon2 paramters")?;
    let hasher = Argon2::new(Algorithm::Argon2id, Version::V0x13, params);
    let password_hash = hasher
        .hash_password(password.0.expose_secret().as_bytes(), &salt)
        .context("Failed to hash password")
        .map_err(AuthError::UnexpectedError)?;

    println!("fdefdf: {:?}", password_hash);
    Ok(PasswordHashPHC::new(password_hash.to_string()))
}
