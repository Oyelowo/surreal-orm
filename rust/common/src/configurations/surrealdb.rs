use super::utils::get_env_vars_by_prefix;
use serde::Deserialize;
use serde_aux::prelude::deserialize_number_from_string;
use surrealdb::Datastore;

#[derive(Deserialize, Debug, Clone)]
#[serde(rename_all = "lowercase")]
pub struct SurrealdbConfigs {
    pub namespace: String,
    pub database: String,
    pub username: String,
    pub password: String,
    pub host: String,

    #[serde(deserialize_with = "deserialize_number_from_string")]
    pub port: u16,

    #[serde(default = "default_require_ssl")]
    pub require_ssl: Option<bool>,
}

fn default_require_ssl() -> Option<bool> {
    Some(false)
}

impl Default for SurrealdbConfigs {
    fn default() -> Self {
        get_env_vars_by_prefix("SURREALDB_")
    }
}

impl SurrealdbConfigs {
    pub fn get_database(self) -> anyhow::Result<Datastore> {
        // let ds = Datastore::new("tikv://127.0.0.1:8000").await?;
        todo!()
    }
}
