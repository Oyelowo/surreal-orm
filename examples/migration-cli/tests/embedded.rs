use std::collections::{BTreeMap, HashMap};
use std::fs;

use surreal_models::migrations::Resources;
use surreal_orm::migrator::{
    self, embed_migrations, DbInfo, Informational, Migration, MigrationConfig, RollbackStrategy,
};
use surreal_orm::statements::{info_for, select};
use surreal_orm::{All, ReturnableSelect, Runnable};
use surrealdb::engine::local::Db;
use surrealdb::engine::remote::ws::Ws;
use surrealdb::opt::auth::Root;
use surrealdb::{Connection, Surreal};
use tempfile::tempdir;

// Embed migrations as constant
const MIGRATIONS_ONE_WAY: migrator::EmbeddedMigrationsOneWay =
    embed_migrations!("tests/migrations-oneway", one_way, strict);

const MIGRATIONS_TWO_WAY: migrator::EmbeddedMigrationsTwoWay =
    embed_migrations!("tests/migrations-twoway", two_way, strict);

#[test]
fn test_embedded() {
    assert_eq!(MIGRATIONS_ONE_WAY.get_migrations().len(), 2);
    assert_eq!(MIGRATIONS_TWO_WAY.get_migrations().len(), 1);

    let migs = MIGRATIONS_ONE_WAY.to_migrations_one_way().unwrap();
    assert_eq!(migs.len(), 2);
    // check the meta data
    assert_eq!(migs[0].name, "20231029202315_create_new_stuff");
    assert_eq!(migs[0].timestamp, 20231029202315);
    insta::assert_snapshot!(migs[0].content);
    assert_eq!(migs[1].name, "20231029224601_create_new_stuff");
    assert_eq!(migs[1].timestamp, 20231029224601);
    assert_eq!(migs.len(), 2);
    assert_eq!(
        migs[1].content,
        "DEFINE FIELD labels ON planet TYPE array;\nUPDATE planet SET labels = tags;\nREMOVE FIELD tags ON TABLE planet;"
    );

    let migs = MIGRATIONS_TWO_WAY.to_migrations_two_way().unwrap();
    assert_eq!(migs.len(), 1);

    // check the meta data
    assert_eq!(migs[0].name, "20231030025711_migration_name_example");
    assert_eq!(migs[0].timestamp, 20231030025711);
    insta::assert_snapshot!(migs[0].up);
    insta::assert_snapshot!(migs[0].down);
}

async fn initialize_db() -> Surreal<impl Connection> {
    // let db = Surreal::new::<Ws>("localhost:8000")
    let db = Surreal::<Db>::new(())
        .await
        .expect("Failed to connect to db");
    // db.signin(Root {
    //     username: "root",
    //     password: "root",
    // })
    // .await
    // .expect("Failed to signin");
    db.use_ns("test").use_db("test").await.unwrap();
    db
}

#[tokio::test]
async fn test_oneway_migrations() {
    let db = initialize_db().await;
    // ONE WAY MIGRATIONS
    let files_config = MigrationConfig::new().make_strict();

    let mig_dir = tempdir().expect("Failed to create temp directory");
    let temp_test_migration_dir = &mig_dir.path().join("migrations-tests");
    let test_migration_name = "test_migration";
    let _files = fs::read_dir(&temp_test_migration_dir).expect_err("Migrations not yet created");

    let one_way = files_config
        .custom_path(&temp_test_migration_dir.display().to_string())
        .one_way();

    let db_info = || async {
        info_for()
            .database()
            .get_data::<DbInfo>(db.clone())
            .await
            .unwrap()
    };

    assert_eq!(
        db_info().await.as_ref().unwrap().tables().get_names(),
        vec![] as Vec<String>
    );

    // Comment out this line to generate oneway migrations
    // To be used from cli
    one_way
        .generate_migrations(test_migration_name, Resources)
        .await
        .unwrap();

    assert_eq!(
        db_info().await.as_ref().unwrap().tables().get_names(),
        vec![] as Vec<String>
    );

    // Run normal non-embedded pending migrations in migration directory
    one_way.run_pending_migrations(db.clone()).await.unwrap();

    // Files would now be created
    let files = fs::read_dir(&temp_test_migration_dir).unwrap();
    assert_eq!(files.count(), 1, "Migration not created");

    assert_eq!(
        db_info().await.unwrap().tables().get_names(),
        vec!["animal", "crop", "eats", "migration", "planet", "student"],
        "Tables not created"
    );

    let mut table_defs = db_info()
        .await
        .unwrap()
        .tables()
        .get_all_definitions()
        .iter()
        .map(|q| q.to_string())
        .collect::<Vec<_>>();

    table_defs.sort();
    insta::assert_snapshot!(table_defs.join("\n"));

    insta::assert_debug_snapshot!(db_info().await.unwrap());
    let fields_info = |field: String| async {
        let fields_info = info_for()
            .table(field)
            .get_data::<BTreeMap<String, BTreeMap<String, String>>>(db.clone())
            .await
            .unwrap();

        insta::assert_snapshot!(fields_info
            .as_ref()
            .unwrap()
            .values()
            .map(|v| v
                .keys()
                .map(|k| k.to_string())
                .collect::<Vec<_>>()
                .join(", "))
            .collect::<Vec<_>>()
            .join("\n"));
        insta::assert_debug_snapshot!(fields_info.unwrap());
    };

    fields_info("animal".to_string()).await;
    fields_info("crop".to_string()).await;
    fields_info("eats".to_string()).await;
    fields_info("migration".to_string()).await;
    fields_info("planet".to_string()).await;
    fields_info("student".to_string()).await;
}

#[tokio::test]
async fn test_twoway_migrations() {
    let db = initialize_db().await;
    let files_config = MigrationConfig::new().make_strict();

    let mig_dir = tempdir().expect("Failed to create temp directory");
    let temp_test_migration_dir = &mig_dir.path().join("migrations-tests");
    let test_migration_name = "test_migration";
    let _files = fs::read_dir(&temp_test_migration_dir).expect_err("Migrations not yet created");

    let two_way = files_config
        .custom_path(&temp_test_migration_dir.display().to_string())
        .two_way();

    let get_db_info = || async {
        info_for()
            .database()
            .get_data::<DbInfo>(db.clone())
            .await
            .unwrap()
    };
    let get_migrations = || async {
        select(All)
            .from(Migration::table_name())
            .return_many::<Migration>(db.clone())
            .await
            .unwrap()
    };

    assert_eq!(
        get_db_info().await.as_ref().unwrap().tables().get_names(),
        vec![] as Vec<String>
    );

    // Comment out this line to generate oneway migrations
    // To be used from cli
    two_way
        .generate_migrations(test_migration_name, Resources)
        .await
        .unwrap();

    assert_eq!(
        get_db_info().await.as_ref().unwrap().tables().get_names(),
        vec![] as Vec<String>
    );

    assert_eq!(get_migrations().await.len(), 0);
    // Run normal non-embedded pending migrations in migration directory
    two_way.run_pending_migrations(db.clone()).await.unwrap();
    assert_eq!(get_migrations().await.len(), 1);
    // assert_eq!(
    //     migrations()
    //         .await
    //         .iter()
    //         .map(|r| r.clone().id)
    //         .collect::<Vec<_>>(),
    //     vec![]
    // );

    // Files would now be created
    let files = fs::read_dir(&temp_test_migration_dir).unwrap();
    assert_eq!(files.count(), 2, "Migration not created");

    assert_eq!(
        get_db_info().await.unwrap().tables().get_names(),
        vec!["animal", "crop", "eats", "migration", "planet", "student"]
    );

    let mut table_defs = get_db_info()
        .await
        .unwrap()
        .tables()
        .get_all_definitions()
        .iter()
        .map(|q| q.to_string())
        .collect::<Vec<_>>();

    table_defs.sort();
    insta::assert_snapshot!(table_defs.join("\n"));

    insta::assert_debug_snapshot!(get_db_info().await.unwrap());
    let fields_info = |field: String| async {
        let fields_info = info_for()
            .table(field)
            .get_data::<BTreeMap<String, BTreeMap<String, String>>>(db.clone())
            .await
            .unwrap();

        insta::assert_snapshot!(fields_info
            .as_ref()
            .unwrap()
            .values()
            .map(|v| v
                .keys()
                .map(|k| k.to_string())
                .collect::<Vec<_>>()
                .join(", "))
            .collect::<Vec<_>>()
            .join("\n"));
        insta::assert_debug_snapshot!(fields_info.unwrap());
    };

    fields_info("animal".to_string()).await;
    fields_info("crop".to_string()).await;
    fields_info("eats".to_string()).await;
    fields_info("migration".to_string()).await;
    fields_info("planet".to_string()).await;
    fields_info("student".to_string()).await;

    two_way
        .rollback_migrations(RollbackStrategy::Latest, db.clone())
        .await
        .unwrap();

    assert_eq!(get_migrations().await.len(), 0);

    // Files would now be deleted
    let files = fs::read_dir(&temp_test_migration_dir).unwrap();
    assert_eq!(files.count(), 0, "Migrations not deleted");
}
