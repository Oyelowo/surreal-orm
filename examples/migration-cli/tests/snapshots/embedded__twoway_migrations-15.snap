---
source: examples/migration-cli/tests/embedded.rs
expression: up_mig
---
[
    "DEFINE TABLE animal SCHEMAFULL;\nDEFINE FIELD updatedAt ON animal TYPE datetime;\nDEFINE FIELD id ON animal TYPE record<animal>;\nDEFINE FIELD createdAt ON animal TYPE datetime;\nDEFINE FIELD species ON animal TYPE string;\nDEFINE FIELD velocity ON animal TYPE int;\nDEFINE FIELD attributes ON animal TYPE array;\nDEFINE INDEX species_speed_idx ON animal FIELDS species, velocity UNIQUE;\nDEFINE EVENT event1 ON animal WHEN species = 'Homo Erectus' AND velocity > 545 THEN (SELECT * FROM crop);\nDEFINE EVENT event2 ON animal WHEN (species = 'Homo Sapien') AND (velocity < 10) THEN (SELECT * FROM eats);\nDEFINE TABLE planet SCHEMAFULL;\nDEFINE FIELD createdAt ON planet TYPE datetime;\nDEFINE FIELD labels ON planet TYPE array;\nDEFINE FIELD firstName ON planet TYPE string;\nDEFINE FIELD population ON planet TYPE int;\nDEFINE FIELD id ON planet TYPE record<planet>;\nDEFINE FIELD updatedAt ON planet TYPE datetime;\nDEFINE TABLE crop SCHEMAFULL;\nDEFINE FIELD id ON crop TYPE record<crop>;\nDEFINE FIELD color ON crop TYPE string;\nDEFINE TABLE student SCHEMAFULL;\nDEFINE FIELD createdAt ON student TYPE datetime;\nDEFINE FIELD age ON student TYPE int;\nDEFINE FIELD updatedAt ON student TYPE datetime;\nDEFINE FIELD id ON student TYPE record<student>;\nDEFINE FIELD university ON student TYPE string;\nDEFINE TABLE migration SCHEMAFULL;\nDEFINE FIELD timestamp ON migration TYPE int;\nDEFINE FIELD name ON migration TYPE string;\nDEFINE FIELD id ON migration TYPE record<migration>;\nDEFINE TABLE eats SCHEMAFULL;\nDEFINE FIELD id ON eats TYPE record<eats>;\nDEFINE FIELD in ON eats TYPE record;\nDEFINE FIELD place ON eats TYPE string;\nDEFINE FIELD out ON eats TYPE record;\nDEFINE FIELD createdAt ON eats TYPE datetime;",
]
